File: akka-sample-sharding-scala/src/main/scala/sample/sharding/Devices.scala
package sample.sharding

import scala.concurrent.duration._
import scala.util.Random

import akka.actor._
import akka.cluster.sharding._

object Devices {
  // Update a random device
  case object UpdateDevice

  case object ReadTemperatures

  case object Start

  def props(): Props =
    Props(new Devices)
}

class Devices extends Actor with ActorLogging with Timers {
  import Devices._

  private val extractEntityId: ShardRegion.ExtractEntityId = {
    case msg @ Device.RecordTemperature(id, _,_, _) => (id.toString, msg)
    case msg @ Device.GetTemperature(id)       => (id.toString, msg)
  }

  private val numberOfShards = 10

  private val extractShardId: ShardRegion.ExtractShardId = {
    case Device.RecordTemperature(id, _, _, _) =>
      (math.abs(id.hashCode) % numberOfShards).toString
    case Device.GetTemperature(id) =>
      (math.abs(id.hashCode) % numberOfShards).toString
    // Needed if you want to use 'remember entities':
    case ShardRegion.StartEntity(id) =>
      (math.abs(id.hashCode) % numberOfShards).toString
  }

  val deviceRegion: ActorRef = ClusterSharding(context.system).start(
    typeName = "Device",
    entityProps = Device.props(),
    settings = ClusterShardingSettings(context.system),
    extractEntityId = extractEntityId,
    extractShardId = extractShardId
  )

  val random = new Random()
  val numberOfDevices = 10

  if (context.system.settings.config.getBoolean("sample.sending-temperatures")) {
    timers.startSingleTimer(Start, Start, 20.seconds)
  }

  private var seqNr = 0L
  private var sequenceNumbers = Map.empty[Long, Int]

  def receive = {
    case Start =>
      timers.startTimerWithFixedDelay(UpdateDevice, UpdateDevice, 100.millis)
      timers.startTimerWithFixedDelay(
        ReadTemperatures,
        ReadTemperatures,
        15.seconds
      )

    case UpdateDevice =>
      seqNr += 1
      //val deviceId = random.nextInt(numberOfDevices)
      val deviceId = (seqNr % numberOfDevices).toInt
      sequenceNumbers = sequenceNumbers.updated(seqNr, deviceId)
      val temperature = 5 + 30 * random.nextDouble()
      val msg = Device.RecordTemperature(deviceId, temperature, System.nanoTime(), seqNr)
      log.info(s"Sending $msg")
      deviceRegion ! msg

    case Device.RecordTemperatureAck(deviceId, startTime, seqNr) =>
      val durationMs = (System.nanoTime() - startTime) / 1000 / 1000
      if (durationMs > 500)
        log.info("Delayed ack of device {} seqNr {} after {} ms", deviceId, seqNr, durationMs)
      else
        log.info("Ack of device {} seqNr {} after {} ms", deviceId, seqNr, durationMs)

      sequenceNumbers -= seqNr
      log.info("Pending sequence numbers: {}", sequenceNumbers)

    case ReadTemperatures =>
      (0 to numberOfDevices).foreach { deviceId =>
        deviceRegion ! Device.GetTemperature(deviceId)
      }

    case temp: Device.Temperature =>
      if (temp.readings > 0)
        log.info(
          "Temperature of device {} is {} with average {} after {} readings",
          temp.deviceId,
          temp.latest,
          temp.average,
          temp.readings
        )
  }
}

File: akka-sample-sharding-java/src/main/java/sample/sharding/Devices.java
package sample.sharding;

import java.time.Duration;
import java.util.Random;
import java.util.concurrent.CompletionStage;

import akka.actor.AbstractActorWithTimers;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.cluster.sharding.ClusterSharding;
import akka.cluster.sharding.ClusterShardingSettings;
import akka.cluster.sharding.ShardRegion;
import akka.pattern.Patterns;

public class Devices extends AbstractActorWithTimers {

  public static Props props() {
    return Props.create(Devices.class, Devices::new);
  }

  private static final int NUMBER_OF_SHARDS = 100;

  static ShardRegion.MessageExtractor messageExtractor =
    new ShardRegion.HashCodeMessageExtractor(NUMBER_OF_SHARDS) {

    @Override
    public String entityId(Object message) {
      if (message instanceof Device.RecordTemperature)
        return String.valueOf(((Device.RecordTemperature) message).deviceId);
      else if (message instanceof Device.GetTemperature)
        return String.valueOf(((Device.GetTemperature) message).deviceId);
      else
        return null;
    }
  };

  public enum UpdateDevice {
    INSTANCE
  }

  public enum ReadTemperatures {
    INSTANCE
  }

  private final LoggingAdapter log = Logging.getLogger(getContext().system(), this);

  private final ActorRef deviceRegion;

  private final Random random = new Random();

  private final int numberOfDevices = 50;

  public Devices() {
    ActorSystem system = getContext().getSystem();
    ClusterShardingSettings settings = ClusterShardingSettings.create(system);
    deviceRegion = ClusterSharding.get(system)
      .start(
        "Device",
        Device.props(),
        settings,
        messageExtractor);

    getTimers().startTimerWithFixedDelay(UpdateDevice.INSTANCE, UpdateDevice.INSTANCE, Duration.ofSeconds(1));
    getTimers().startTimerWithFixedDelay(ReadTemperatures.INSTANCE, ReadTemperatures.INSTANCE, Duration.ofSeconds(15));
  }

  @Override
  public Receive createReceive() {
    return receiveBuilder()
      .match(UpdateDevice.class, m -> receiveUpdateDevice())
      .match(ReadTemperatures.class, m -> receiveReadTemperatures())
      .match(Device.Temperature.class, this::receiveTemperature)
      .build();
  }

  private void receiveUpdateDevice() {
    Integer deviceId = random.nextInt(numberOfDevices);
    Double temperature = 5 + 30 * random.nextDouble();
    Device.RecordTemperature msg = new Device.RecordTemperature(deviceId, temperature);
    log.info("Sending {}", msg);
    deviceRegion.tell(msg, getSelf());
  }

  private void receiveReadTemperatures() {
    for (int deviceId = 0; deviceId < numberOfDevices; deviceId++) {
      if (deviceId >= 40) {
        CompletionStage<Object> reply = Patterns.ask(deviceRegion, new Device.GetTemperature(deviceId), Duration.ofSeconds(3));
        Patterns.pipe(reply, getContext().getDispatcher()).to(getSelf());
      } else {
        deviceRegion.tell(new Device.GetTemperature(deviceId), getSelf());
      }
    }
  }

  private void receiveTemperature(Device.Temperature temp) {
    if (temp.readings > 0)
      log.info(
        "Temperature of device {} is {} with average {} after {} readings",
        temp.deviceId,
        temp.latest,
        temp.average,
        temp.readings
      );
  }

}
